### ISSUE TITLE
Request change to IRS2 refpix reference file format and corresponding IRS2 refpix correction step

### ISSUE THREAD

**Issue by stscijgbot-jp** [spacetelescope/jwst#8192]
_Issue [JP-3508](https://jira.stsci.edu/browse/JP-3508) was created on JIRA by [Eddie Bergeron](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bergeron):_

There are some issues with the current NIRSpec irs2 refpix reference files. The current reference files have the following format:

#########################################################

IDL> fits_info,'█████████████████████████████████████████████████████████████
████████████████████████████████████████████████████████████ has 5 extensions
Primary header: 50 records
No data
Extension 1 – IRS2    
               Header : 28 records
               Binary Table ( 32 2916352 )
Extension 2 – IRS2ALPHA
               Header : 19 records
               Binary Table ( 16 2916352 )
Extension 3 – IRS2BETA
               Header : 19 records
               Binary Table ( 16 2916352 )
Extension 4 – DQ      
               Header : 19 records
               Binary Table ( 6 8 )
Extension 5 – ASDF    
               Header : 12 records
               Binary Table ( 2455 1 )
 

IDL> print,tag_names(b)
ALPHA_0 ALPHA_1 ALPHA_2 ALPHA_3 BETA_0 BETA_1 BETA_2 BETA_3
IDL> print,tag_names(c)
ALPHA_0 ALPHA_1 ALPHA_2 ALPHA_3
IDL> print,tag_names(d)
BETA_0 BETA_1 BETA_2 BETA_3
IDL> print,tag_names(e)
OUTPUT ODD_EVEN MASK
IDL> print,tag_names(f)
ASDF_METADATA

#########################################################

I spoke with Stephan Birkmann (original author of this reference file) about the format. The first extension here "IRS2" contains both the alpha and beta coefficients that the pipeline is currently using. However the second and third extensions (IRS2ALPHA, IRS2BETA) also contain separate but different alpha and beta coeffs. Stephan said these are test coeffs they created before launch to test alpha-only and beta-only corrections. These are not needed and should be removed from the reference file to avoid confusion. If single-coeff solutions are desired in the future this should probably be handled with a different reference file, and likely can be handled with the single "IRS2" extension containing both coeffs by simply setting the undesired alpha or beta coeffs to 0's. This also has the advantage that it should work with no pipeline code changes to accommodate alpha or beta-only corrections.

Additionally, the naming convention of "alpha" and "beta" coefficients in the reference file are reversed relative to the published PASP reference on the irs2 correction and the IDL code provided with that reference for creating the alpha and beta coeffs in the first place. The pipeline python code has comments to the effect that they are reversed, so i{*}t is working correctly{*} with the existing reference file, but it would be better all around if both the reference file and the pipeline code were changed to match the sense in the original reference. That is, the alpha coefficients should apply to the refout channel and the beta coefficients should apply to the interleaved reference pixels. Again, the pipeline is currently working correctly with the existing reference file, but the convention is reversed relative to the published reference and IDL source code and this might lead to confusion for anyone attempting to create new coeffs in the future.

Finally, after discussions with Stephan, Melanie and Mike Regan, we feel that the use of the bitmapped DQ extension in this reference file is needlessly complex just to be compact. It would be easier to maintain if this DQ extension was a simple 2048-element wide table or vector with badpix DQ flags - such as a single line from the regular static bad pixel mask. Better still, it might be better to eliminate this irs2-specific DQ extension from this reference file and simply modify the pipeline code to look at the appropriate reference row line in the static bad pixel mask (currently the 3rd row from the bottom of the 2048x2048 array). This way any updates to the static mask would automatically carry through to the irs2 correction without the need to update the irs2 refpix reference file at the same time. The irs2 refpix reference file would then contain only the "IRS2" extension with both the alpha and beta coeffs, and the ASDF extension and nothing else. Should consult with the NIRSpec team on these suggestions.

 

In summary, I'm recommending the following changes to the format of the NIRSpec irs2 refpix reference files:

_1) remove unused "extra" coeff tables and just have one table in the reference file. I suspect that we could achieve single coeff correction simply by setting one or the other set of coeffs to 0's and the pipeline code as-written would then zero-weight those automatically._

*2) reverse the sense of the alpha and beta coeffs in both the reference file and take out the corresponding reversal in the pipeline code to match*

_3a) change the DQ extension to be a 2048-wide "bad pixel mask" vector. This could be a fits table of DQ flags or a [2048] wide "image" in that extension. Ideally this would just be a copy of the corresponding line from the regular NRS bad pixel masks._ 

            __            

_-- or --_

_
3b) alternately we could just leave out the DQ extension altogether and add logic to the pipeline code to read the regular static bad pixel map and handle as desired based on those flags. This would be my preferred choice, since there would be no need to modify this reference file when new bad pixel maps are made in the future. Note that if the OTF flagging is working in all cases this DQ step becomes irrelevant anyways._

**Comment by stscijgbot-jp** [comment]
_Comment by [Eddie Bergeron](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bergeron) on [JIRA](https://jira.stsci.edu/browse/JP-3508?focusedCommentId=679384#comment-679384):_

For reference, here is the Rauscher 2017 PASP reference for the irs2 correction:

[^Rauscher_2017_PASP_129_105003_irs2.pdf]

**Comment by stscijgbot-jp** [comment]
_Comment by [Melanie Clarke](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mclarke) on [JIRA](https://jira.stsci.edu/browse/JP-3508?focusedCommentId=680964#comment-680964):_

The NIRSpec team will need to discuss the impact these changes would have on reference file generation and processing before we can proceed.

**Comment by stscijgbot-jp** [comment]
_Comment by [David Law](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dlaw) on [JIRA](https://jira.stsci.edu/browse/JP-3508?focusedCommentId=705194#comment-705194):_

Assigning to [Christian Hayes](https://jira.stsci.edu/secure/ViewProfile.jspa?name=chayes) to discuss within the NIRSpec team.

**Comment by stscijgbot-jp** [comment]
_Comment by [David Law](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dlaw) on [JIRA](https://jira.stsci.edu/browse/JP-3508?focusedCommentId=983417#comment-983417):_

Any update [Christian Hayes](https://jira.stsci.edu/secure/ViewProfile.jspa?name=chayes) ?

**Comment by stscijgbot-jp** [comment]
_Comment by [Christian Hayes](https://jira.stsci.edu/secure/ViewProfile.jspa?name=chayes) on [JIRA](https://jira.stsci.edu/browse/JP-3508?focusedCommentId=983905#comment-983905):_

No updates.  I'm putting this on hold for now since it was a lower priority last time we discussed calibration file updates, but tagging [Elaine Mae Frazer](https://jira.stsci.edu/secure/ViewProfile.jspa?name=efrazer) [Peter Zeidler](https://jira.stsci.edu/secure/ViewProfile.jspa?name=zeidler) [James Muzerolle](https://jira.stsci.edu/secure/ViewProfile.jspa?name=muzerol), to put this on your all's radar as calibration reference file updates we need to look into making and testing.

**Comment by stscijirahub** [comment]
_This issue is tracked on JIRA as [JP-3508](https://jira.stsci.edu/browse/JP-3508)._
